finat/hdivcurl.py

```diff
@@ -70,6 +70,14 @@ def point_evaluation(self, order, refcoords, entity=None):
         core_eval = self.wrappee.point_evaluation(order, refcoords, entity)
         return self._transform_evaluation(core_eval)
 
+    @property
+    def dual_basis(self):
+        Q, x = self.wrappee.dual_basis
+        beta = self.get_indices()
+        zeta = self.get_value_indices()
+        Q = gem.ListTensor(self.transform(gem.partial_indexed(Q, beta)))
+        return gem.ComponentTensor(Q[zeta], beta + zeta), x
```

Comment:
This needs some explanatory comments (along the lines of what you have written in [this comment](https://github.com/FInAT/FInAT/pull/98#issuecomment-913717573).

I'm also surprised to see a `gem.ListTensor` and `gem.partial_indexed` appear here. What's going on on that line?

See also the other comment about beta.

Reply:
`Q` has shape `(*self.index_shape, *wrappee.value_shape)`, `transform` promotes something of `wrappee.value_shape` to a `list` of `self.value_shape` which is then `ListTensor`ed. So now we have `Q` with shape `self.value_shape` (good) and free indices `beta`, which we need to turn into a componenttensor.

Reply:
Okay so tell me if this is correct:

```
gem.partial_indexed(Q, beta))
```
makes sure your wrappe element's `Q` only has shape `wrappee.value_shape` which is then transformed with `self.transform`.

After being `self.transform`ed we now have a `self.value_shape` long list of tensors which need to be stacked (i.e. wrapped in a `gem.ListTensor`) to get a new `Q`.
This new `Q` has all the right content for our new element (though I'm not sure what that content is in this case).
 - `Q.shape` is `zeta` (?)
 - `Q` has any number of free indices

Lastly we transform our `Q` to match the API by indexing out the shape then re-adding it such that we keep x.indices as our free indices.

Reply:
> Okay so tell me if this is correct:
> 
> ```
> gem.partial_indexed(Q, beta))
> ```
> 
> makes sure your wrappe element's `Q` only has shape `wrappee.value_shape` which is then transformed with `self.transform`.
> 
> After being `self.transform`ed we now have a `self.value_shape` long list of tensors which need to be stacked (i.e. wrapped in a `gem.ListTensor`) to get a new `Q`.
> This new `Q` has all the right content for our new element (though I'm not sure what that content is in this case).

Think of an example where we're hdiving a tensor product of P x DP on a quadrilateral. Each of the wrappee element functionals are scalar point evaluations, the hdiv functionals are just these scalar evaluations promoted to be vectors dotted with an appropriate normal. Since we're on the reference cell and know what the normal is, this just turns the functional into a vector `[0, -v]` (evaluate the second component of the vector multiply by minus-one).

The scalar functional might just be something like point evaluation, but the transformed functional is point evaluation of `-1 * third component`, so the initial functional is turned into `[0, 0, -v]`

> 
> * `Q.shape` is `zeta` (i.e. self.value_shape) (?)

Yes.

> * `Q` has any number of free indices though we know that they include `beta` (the indices along the basis)

By design, it must have beta + x.indices, I think.

> Lastly we transform our `Q` to match the API by indexing out the shape then re-adding it such that we keep `x.indices` as our free indices.

Yes.

Reply:
Okay good. If you can add some comments here saying (briefly) what you've put here then I think this can be resolved

Reply:
Added a short comment.